File: tools/agentic-starter-kit/python/src/webex_client.py
"""Small Webex API client used by the starter examples.

The methods intentionally stay simple so beginners can follow the request flow
without needing a framework or SDK.
"""

import os
from typing import Any, Dict, List, Optional

import requests
from dotenv import load_dotenv


load_dotenv()


class WebexClient:
    def __init__(self) -> None:
        self.access_token = os.getenv("WEBEX_ACCESS_TOKEN")
        self.base_url = os.getenv("WEBEX_API_BASE_URL", "https://webexapis.com/v1").rstrip("/")
        self.max_page_size = int(os.getenv("WEBEX_MAX_PAGE_SIZE", "50"))

    def _require_token(self) -> str:
        if not self.access_token:
            raise ValueError("Missing required environment variable: WEBEX_ACCESS_TOKEN")
        return self.access_token

    def _headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self._require_token()}",
            "Content-Type": "application/json",
        }

    def _get_all_pages(self, path: str, params: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        next_url = f"{self.base_url}{path}"
        request_params = dict(params or {})
        request_params.setdefault("max", self.max_page_size)
        items: List[Dict[str, Any]] = []

        while next_url:
            response = requests.get(next_url, headers=self._headers(), params=request_params, timeout=60)
            response.raise_for_status()
            payload = response.json()
            items.extend(payload.get("items", []))
            next_url = response.links.get("next", {}).get("url")
            # Follow-up pages already contain the query string in the next link.
            request_params = {}

        return items

    def list_spaces(self) -> List[Dict[str, Any]]:
        # The Webex API path is /rooms even though users usually say "spaces".
        return self._get_all_pages("/rooms")

    def list_messages(self, room_id: str) -> List[Dict[str, Any]]:
        return self._get_all_pages("/messages", {"roomId": room_id})

    def list_recordings(self) -> List[Dict[str, Any]]:
        return self._get_all_pages("/recordings")

    def list_transcripts(self) -> List[Dict[str, Any]]:
        return self._get_all_pages("/meeting/transcripts")

    def get_transcript_details(self, transcript_id: str) -> Dict[str, Any]:
        response = requests.get(
            f"{self.base_url}/meeting/transcripts/{transcript_id}",
            headers=self._headers(),
            timeout=60,
        )
        response.raise_for_status()
        return response.json()

File: tools/agentic-starter-kit/node/src/webexClient.js
import dotenv from "dotenv";

dotenv.config({ path: new URL("../../.env", import.meta.url) });

export class WebexClient {
  constructor() {
    this.accessToken = process.env.WEBEX_ACCESS_TOKEN;
    this.baseUrl = (process.env.WEBEX_API_BASE_URL || "https://webexapis.com/v1").replace(/\/$/, "");
    this.maxPageSize = Number(process.env.WEBEX_MAX_PAGE_SIZE || 50);
  }

  requireToken() {
    if (!this.accessToken) {
      throw new Error("Missing required environment variable: WEBEX_ACCESS_TOKEN");
    }
    return this.accessToken;
  }

  buildHeaders() {
    return {
      Authorization: `Bearer ${this.requireToken()}`,
      "Content-Type": "application/json"
    };
  }

  async getAllPages(path, params = {}) {
    const items = [];
    let nextUrl = new URL(`${this.baseUrl}${path}`);

    for (const [key, value] of Object.entries({ max: this.maxPageSize, ...params })) {
      nextUrl.searchParams.set(key, String(value));
    }

    while (nextUrl) {
      const response = await fetch(nextUrl, { headers: this.buildHeaders() });
      if (!response.ok) {
        throw new Error(`Webex request failed with status ${response.status}`);
      }

      const json = await response.json();
      items.push(...(json.items || []));

      // Webex pagination uses an HTTP Link header instead of a body field.
      const linkHeader = response.headers.get("link") || "";
      const nextMatch = linkHeader.match(/<([^>]+)>;\s*rel="next"/i);
      nextUrl = nextMatch ? new URL(nextMatch[1]) : null;
    }

    return items;
  }

  listSpaces() {
    // The API path is /rooms, but most users know these as spaces.
    return this.getAllPages("/rooms");
  }

  listMessages(roomId) {
    return this.getAllPages("/messages", { roomId });
  }

  listRecordings() {
    return this.getAllPages("/recordings");
  }

  listTranscripts() {
    return this.getAllPages("/meeting/transcripts");
  }

  async getTranscriptDetails(transcriptId) {
    const response = await fetch(`${this.baseUrl}/meeting/transcripts/${transcriptId}`, {
      headers: this.buildHeaders()
    });

    if (!response.ok) {
      throw new Error(`Transcript request failed with status ${response.status}`);
    }

    return response.json();
  }
}
